### ISSUE TITLE
ndbm can't iterate through values on OS X

### ISSUE THREAD

**Issue by d4363c04-2bcc-4b04-b2a4-f8d49c85bdc0** [python/cpython#74573]
BPO | [30388](https://bugs.python.org/issue30388)
--- | :---
Nosy | @terryjreedy, @ronaldoussoren, @ned-deily, @emilyemorehouse
Files | <li>[ndbm.zip](https://bugs.python.org/file46872/ndbm.zip "Uploaded as application/zip at 2017-05-17.16:03:54 by Forest Gregg"): The ndbm files created by shelve.</li><li>[ndbm-repro.c](https://bugs.python.org/file49627/ndbm-repro.c "Uploaded as text/plain at 2020-11-27.11:32:30 by @ronaldoussoren")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-05-17.16:03:55.042>
labels = ['OS-mac', 'type-bug', 'library']
title = "ndbm can't iterate through values on OS X"
updated_at = <Date 2020-11-27.11:32:30.539>
user = 'https://bugs.python.org/ForestGregg'
```

bugs.python.org fields:
```python
activity = <Date 2020-11-27.11:32:30.539>
actor = 'ronaldoussoren'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)', 'macOS']
creation = <Date 2017-05-17.16:03:55.042>
creator = 'Forest Gregg'
dependencies = []
files = ['46872', '49627']
hgrepos = []
issue_num = 30388
keywords = []
message_count = 11.0
messages = ['293858', '293934', '293974', '293975', '294040', '295189', '295215', '295282', '295512', '380963', '381944']
nosy_count = 5.0
nosy_names = ['terry.reedy', 'ronaldoussoren', 'ned.deily', 'emilyemorehouse', 'Forest Gregg']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue30388'
versions = ['Python 3.5']
```

</p></details>

### Linked PRs
* gh-113354
* gh-113431
* gh-113432

**Comment by d4363c04-2bcc-4b04-b2a4-f8d49c85bdc0** [comment]
On Mac OS 10.12.4, a large shelve, backed by ndbm, can be created. But when I attempt to iterate through the values of the shelve it raises this exception:


  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/dedupe/api.py", line 281, in \_blockData
    for block in viewvalues(blocks):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/_collections_abc.py", line 693, in \_\_iter__
    for key in self.\_mapping:
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/shelve.py", line 95, in \_\_iter__
    for k in self.dict.keys():
SystemError: Negative size passed to PyBytes_FromStringAndSize

I've confirmed that this works with Python 3.6.1. All the Python versions were installed from homebrew.

I cannot reproduce on linux on windows machines. I've attached a zip file of the ndbm files that python created.

**Comment by d4363c04-2bcc-4b04-b2a4-f8d49c85bdc0** [comment]
The ndbm db's two files (in the attachment) have the following size

tmp___otctx 0 bytes
tmp___otctx.db 12857344 bytes

**Comment by terryjreedy** [comment]
Do you mean that on OSX, you code works with 3.6.1 and fails with 3.5.3?

**Comment by d4363c04-2bcc-4b04-b2a4-f8d49c85bdc0** [comment]
Very sorry for the ambiguity. 

The bug appears for both python 3.5 and python 3.6.1 on OS X. I have not tried other versions of python on OS X.

**Comment by ned-deily** [comment]
Can you provide a complete example of how to reproduce the behavior you are seeing?  Ideally, it would include producing the shelve object.  Also, please provide the output of this command:

otool -L $(python3.6 -c 'import \_dbm;print(_dbm.\_\_file__)')

**Comment by d4363c04-2bcc-4b04-b2a4-f8d49c85bdc0** [comment]
I have been trying to make a small reproducible example, but haven't been able to isolate it. 

Running this script [1] csv_example on mac os x under either py27 or py3 does seem \*often* cause this problem [2], [3]:

[1] https://github.com/dedupeio/dedupe-examples/blob/master/csv_example/csv_example.py
[2] https://github.com/dedupeio/dedupe/issues/571
[3] https://github.com/dedupeio/dedupe-examples/issues/54

**Comment by ned-deily** [comment]
It would be helpful if you could provide the output of this command:

otool -L $(python3.6 -c 'import \_dbm;print(_dbm.\_\_file__)')

if necessary, substituting whatever command name(s) you are using to invoke python when failing.

**Comment by d4363c04-2bcc-4b04-b2a4-f8d49c85bdc0** [comment]
From one user who had problems under both 3.5 and 3.6

otool -L $(python3.5 -c 'import \_dbm;print(_dbm.\_\_file__)')

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/lib-dynload/_dbm.cpython-35m-darwin.so:
/usr/lib/libSystem.B.dylib (compatibility version 1.0.0,
current version 125.2.0)

otool -L $(python3.6 -c 'import \_dbm;print(_dbm.\_\_file__)')

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/lib-dynload/_
dbm.cpython-36m-darwin.so:

/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version
125.2.0)

**Comment by d4363c04-2bcc-4b04-b2a4-f8d49c85bdc0** [comment]
A different user:

otool -L $(python3.6 -c 'import \_dbm;print(_dbm.file)')
/usr/local/var/pyenv/versions/3.6.1/lib/python3.6/lib-dynload/_dbm.cpython-36m-darwin.so:
/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.50.2)

**Comment by ronaldoussoren** [comment]
The best we can do is document the issue.

**Comment by ronaldoussoren** [comment]
The just merged PRs document the problem in main, 3.12 and 3.11.  As we cannot fix Apple's code that's the best we can do.
